internal/biz/model/relationships.go

```diff
@@ -43,3 +45,28 @@ func (data RelationshipReporter) Value() (driver.Value, error) {
 func (data *RelationshipReporter) Scan(value interface{}) error {
 	return Scan(value, data)
 }
+
+func (r *Relationship) BeforeCreate(db *gorm.DB) error {
+	var err error
+	if r.ID == uuid.Nil {
+		r.ID, err = uuid.NewV7()
+		if err != nil {
+			return fmt.Errorf("failed to generate resource uuid: %w", err)
+		}
+	}
+
+	if r.SubjectId == uuid.Nil {
```

Comment:
Based on my super naive understanding of relationships I believe subject and object should already exist before the relationship is able to be created, is that correct? If so, we can definitely remove the SubjectId and ObjectId fallback uuid creation here since we won't need to ever create these when a relationship is created.

Reply:
That's right - these should exist already. 
+1 on removing the SubjectId and ObjectId checks

Reply:
my also naive understanding here; subject ID and object ID in this scenario are unique inside of the inventory DB are are not configurable/exposed via API. 

When a user creates a relationship they DO have to provide a subject_local_resource_id and object_local_resource_id which are set in the reporter_data section of the request/table. My understanding is those local_resource_id values are supposed to be some sort of ID that is unique to the requesting system (think OCM clusters to provision shards or whatever values)

The subject_id and object_id values were made UUID to improve queries in inventory as they are used to find a relationship based on https://github.com/project-kessel/inventory-api/blob/a85a24d48f9b4ce3130a4f06c7d6dd8cc12e82d0/internal/data/relationships/relationships.go#L105

https://github.com/project-kessel/docs/blob/main/src/content/docs/inventory/inventory-api.md#reporter_data-1

Reply:
sample relationship in DB where subjectID and objectID are UUID and the provided local resource ids show in reporter section

```kessel-inventory=# select * from relationships;
-[ RECORD 1 ]-----+----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
id                | 0192fe0f-3520-7e45-bf91-e6396d872872
org_id            | 
relationship_data | {"status": 3}
relationship_type | k8s-policy_is-propagated-to_k8s-cluster
subject_id        | 0192fe0f-095f-7be6-a25a-a15eab7ffc20
object_id         | 0192fe0e-c7e9-7cb1-9c12-02bc105d5b40
reporter          | {"reporter_id": "curl/8.6.0", "reporter_type": "OCM", "reporter_version": "0.1", "object_resource_type": "k8s-cluster", "subject_resource_type": "k8s-policy", "object_local_resource_id": "2", "subject_local_resource_id": "1"}
created_at        | 
updated_at        | 2024-11-05 20:41:15.727725+00
```

Reply:
In the case of relationships, when we are creating one, we do get a subject_local_resource_id and object_local_resource_id.
These values (along with the types and resource data) should map to `resource` (which already have an id) - we take the id from there and populate or model.

See
https://github.com/project-kessel/inventory-api/blob/main/internal/biz/relationships/relationships.go#L48-L66

Reply:
I'm following now. Just updated the logic and tested it and it makes perfect sense. Good eyes @tylercreller !